### PR TITLE
Fix lookups of streams to include architecture

### LINKED ISSUE #212: libmodulemd 2.x doesn't handle repodata where there are multiple stream documents for separate arches of the same NSVC
This may require a major-version soname bump. Try to avoid this if at all possible.

### PR DESCRIPTION
**Pull request by sgallagher**
Fixes #212 

The first two patches in the series are minor bugfixes noticed while working on the other bits.

The third patch moves the getter and setter for "arch" to `Modulemd.ModuleStream` rather than `Modulemd.ModuleStreamV1` and Modulemd.ModuleStreamV2`, because that information will be needed at this layer to support the `Modulemd.ModuleStream.get_NSVCA()` addition.

The fourth patcch introduces the new method `Modulemd.ModuleStream.get_NSVCA()`. It deprecates the previous function `Modulemd.ModuleStream.get_nsvc()`, generating non-fatal warnings during compilation and Python execution. It also modifies the `get_nsvc()` function to be a wrapper around  `Modulemd.ModuleStream.get_NSVCA()`.

The fifth patch adds a new search method: `Modulemd.Module.get_stream_by_NSVCA()`, which deprecates ` Modulemd.Module.get_stream_by_nsvc()` in favor of the version supporting architecture.

The sixth and final patch adds support to `Modulemd.Module.add_stream()` to consider architecture in addition to name, stream, version and context when determining if a particular stream is already present in the `Modulemd.Module`. Originally, I had planned to make it an error when the same NSVCA is encountered, but while writing this, I realized that the `Modulemd.ModuleIndexMerger` relies on the current implementation of deduplicating them by removing the older one. I added a TODO to the code for throwing an error if we determine that they are not in fact exactly identical, but that is currently blocked by #186.

### CODE REVIEW

**Comment by sgallagher** [comment]
While filing this, I realized I don't actually have the correct deprecations around `Modulemd.ModuleStream.get_nsvc()`, so I'll fix that quickly and then force-push.

**Comment by sgallagher** [comment]
The missing deprecations are now present.

File: modulemd/v2/include/modulemd-2.0/modulemd-module-stream.h [hunk 1]
```diff
@@ -352,11 +380,28 @@ modulemd_module_stream_get_context (ModulemdModuleStream *self);
...
 modulemd_module_stream_get_nsvc_as_string (ModulemdModuleStream *self);
 
 
+/**
+ * modulemd_module_stream_get_NSVCA_as_string: (rename-to modulemd_module_stream_get_NSVCA)
+ * @self: (in): This #ModulemdModuleStream.
+ *
+ * Returns: (transfer full): The NSVCA of this module stream. The returned
+ * format is described
+ * [here](https://docs.fedoraproject.org/en-US/modularity/architecture/consuming/naming-policy/#_forms).
```
Comment (Conan-Kudo) [inline modulemd/v2/include/modulemd-2.0/modulemd-module-stream.h]: Will this reference work when generating the documentation? I don't recall if `gtk-doc` actually handles Markdown.

File: modulemd/v2/modulemd-module-index.c [hunk 2]
```diff
@@ -353,9 +353,9 @@ compare_stream_SVC (gconstpointer a, gconstpointer b)
 {
   ModulemdModuleStream *a_ = *(ModulemdModuleStream **)a;
   ModulemdModuleStream *b_ = *(ModulemdModuleStream **)b;
-  g_autofree gchar *a_nsvc = modulemd_module_stream_get_nsvc_as_string (a_);
-  g_autofree gchar *b_nsvc = modulemd_module_stream_get_nsvc_as_string (b_);
-  return g_strcmp0 (a_nsvc, b_nsvc);
+  g_autofree gchar *a_nsvca = modulemd_module_stream_get_NSVCA_as_string (a_);
+  g_autofree gchar *b_nsvca = modulemd_module_stream_get_NSVCA_as_string (b_);
+  return g_strcmp0 (a_nsvca, b_nsvca);
```
Comment (Conan-Kudo) [inline modulemd/v2/modulemd-module-index.c]: Is it a good idea to do it this way? Shouldn't we have this deprecated for an explicitly SVCA variant?

File: modulemd/v2/include/modulemd-2.0/modulemd-module-stream.h (hunk 1 above)
Comment (sgallagher) [reply modulemd/v2/include/modulemd-2.0/modulemd-module-stream.h]: https://developer.gnome.org/gtk-doc-manual/stable/documenting_syntax.html.en

> As of version 1.20, the preferred way is to use a subset of the basic text formatting syntax called Markdown. On older GTK-Doc versions any documentation that includes Markdown will be rendered as is.

So this should work fine on any platform we're likely to care about.

File: modulemd/v2/modulemd-module-index.c (hunk 2 above)
Comment (sgallagher) [reply modulemd/v2/modulemd-module-index.c]: This is a static, internal function. I can rename it to `compare_stream_SVCA()` if you prefer. It's called in exactly one place, as a comparison function for a sort.

Comment (Conan-Kudo) [reply modulemd/v2/modulemd-module-index.c]: I'd prefer us to be explicit, even with the function definition for an internal function. :)

Comment (sgallagher) [reply modulemd/v2/modulemd-module-index.c]: Fixed and pushed.

**Pull request merged**